---
title: "genotype-preprocess"
output: html_document
date: "2023-04-18"
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Aims

-   Preprocess phased genome information from TOPMed WHI samples
-   Get allele specific and total specific read counts per gene
-   Prepare data format for Cibersort cell imputation

#### Cohort

-   \~ 1350 female African American participants aged between 50-78
    years old, a subset of the Women's Health Initiative (WHI) cohort.
-   Whole genome sequencing (WGS) and whole blood RNA-seq data is
    available for all participants
-   Sampling for RNA-sequencing was conducted at the Long Life Study
    (LLS) time point

#### Data locations

Phased WGS [data:\\](data:\){.uri}
[sftp://mjohnso5\@rhino/fh/scratch/delete90/kooperberg_c/topmed_freeze10/phased/whi_only](sftp://mjohnso5@rhino/fh/scratch/delete90/kooperberg_c/sct_rnaseq/release_files){.uri}

RNA seq data:

BAM files\
[sftp://mjohnso5\@rhino/fh/scratch/delete90/kooperberg_c/sct_rnaseq/release_files](sftp://mjohnso5@rhino/fh/scratch/delete90/kooperberg_c/sct_rnaseq/release_files){.uri}\
Mapped read .gct files:\
\~/Documents/whi_sca/rna/whi_topmed_to6_rnaseq_gene_reads.gct.gz

## RNA pre-processing

Pipeline based on: <https://github.com/pllittle/CSeQTLworkflow>

-   Reads are already mapped to genes saved as .gct files
-   Data needs to be in .bam format for pipline
-   Pipeline github has steps for alignment/qc of raw BAM files, however
    this has already been performed
-   Therefore assign appropriate directories and proceed from
    allele-specific and total read count mapping stage

**Read Processing/Alignment:**

Our processing pipeline consists of the following elements:

1.  Base calls generated in real-time on the NovaSeq6000 instrument (RTA
    3.1.5).

2.  Demultiplexed, unaligned BAM files produced by Picard
    ExtractIlluminaBarcodes and IlluminaBasecallsToSam are converted to
    FASTQ format using SamTools bam2fq (v1.4).

3.  Sequence read and base quality are checked using the FASTX-toolkit
    (v0.0.13).

4.  Sequences are aligned to GRCh38 with reference transcriptome GENCODE
    release 30 using STAR (v2.6.1d).

**File formats:**

-   *accepted_hits.merged.markeddups.recal.bam, .bai [STAR]*\
    This is a BAM file that holds all the read-data and the alignments
    for reads mapped onto the genome including chromosomes,
    mitochondria, spike-in targets, and un-assigned contigs. This file
    holds all the read-data and can be converted back to fastq for
    realignment if desired. The .bai file is the index for the genome
    file that is required for random access by samtools and other
    programs. \<- use this

*transcriptome_hits.merged.bam [RSEM]*\
RSEM generates a BAM file for the alignments to the transcriptome but
the sequence sources are the complete annotated collection of
transcripts (not chromosomes). This file holds the sample alignments in
BAM format to all known transcripts.

```{r settings}

knitr::opts_chunk$set(echo = TRUE, eval = TRUE, warning = FALSE, message = FALSE)
knitr::opts_knit$set(root.dir = '~/Documents/whi_sca/rna')

```

**Generate total, and allele-specific read counts:**

Download asSeq source package

```{bash 1-aseq}

 url=https://github.com/Sun-lab/asSeq/raw
 url=$url/master/asSeq_0.99.501.tar.gz
 
 wget $url

```

```{r install-aseq}

install.packages(pkgs = "asSeq_0.99.501.tar.gz",
 	type = "source",repos = NULL)
 
 PE = TRUE 
 	# set TRUE for paired-end samples
 	# set FALSE for single-end
 
 flag1 = Rsamtools::scanBamFlag(isUnmappedQuery = FALSE,
 	isSecondaryAlignment = FALSE,isDuplicate = FALSE,
 	isNotPassingQualityControls = FALSE,
 	isSupplementaryAlignment = FALSE,isProperPair = PE)
 
 param1 = Rsamtools::ScanBamParam(flag = flag1,
 	what = "seq",mapqFilter = 255)
 
 bam_file = "output_hg38.sortedByCoordinate.md.bam"
 bam_filt_fn = "output.filtered.asSeq.bam"
 Rsamtools::filterBam(file = bam_file,
 	destination = bam_filt_fn,
 	param = param1)

```

Create exon image file:

ftp gencode file

```{r exon image}


 gtf_fn = "gencode.v26.GRCh38.ERCC.genes.gtf.gz"
 exdb = GenomicFeatures::makeTxDbFromGFF(file = gtf_fn,
 	format = "gtf")
 exons_list_per_gene = GenomicFeatures::exonsBy(exdb,
 	by = "gene")
 
 gtf_rds_fn = "exon_by_genes_gencode_v26.rds"
 saveRDS(exons_list_per_gene, gtf_rds_fn)
```
